Change add_noise to match the target variance by scaling by its sqrt.

Rupt opened the pull request:
Resolve issue #281.

Add a test for this noise scaling which fails prior to this change.

Rupt commented:
Edited to avoid sqrt after type casting.

mkunesch commented on optax/_src/transform.py:
```diff
@@ -717,12 +717,13 @@ def update_fn(updates, state, params=None):  # pylint: disable=missing-docstring
     treedef = jax.tree_structure(updates)
     count_inc = numerics.safe_int32_increment(state.count)
     variance = eta / count_inc**gamma
+    std = jnp.sqrt(variance)
```
Nit: std is a standard abbreviation so it's perfectly fine as is, but I'd personally prefer `standard_deviation`, also to be consistent with the use of `variance` over `var`.

Rupt replied:
Done.
(I chose std to match its use in stochastic_gradient_estimators, but totally agree with your argument)

mkunesch commented on optax/_src/transform_test.py:
```diff
@@ -202,6 +202,35 @@ def test_centralize(self, inputs, outputs):
+      scale = jnp.sqrt(eta / i**gamma)
+
+      def rescale(g):
+        return g * scale  # pylint:disable=cell-var-from-loop
```
I think this is simple enough to be written as lambda in the line below. That way you also wouldn't need to disable the pylint warning or did pylint complain about something else then?

Rupt replied:
Yes this is because pylint complains with the lambda form, and with this same warning. This style copies [L186](https://github.com/deepmind/optax/blob/a0d5ed5539a58cc0bfa1d5948c87d2aa3b693d92/optax/_src/transform_test.py#L186).

Rupt replied:
Neat solution, done with the lambda.

mkunesch commented on optax/_src/transform_test.py:
```diff
@@ -202,6 +202,35 @@ def test_centralize(self, inputs, outputs):
+    state_unit = noise_unit.init(params)
+
+    # Check the noise itself by adding it to zeros.
+    updates = jax.tree_map(jnp.zeros_like, self.per_step_updates)
```
Nit: you could make this more self-contained by replacing `self.per_step_updates` with `params` directly (because that's ultimately the shape we want to match).

Rupt replied:
Nice! Done.

mkunesch commented on optax/_src/transform_test.py:
```diff
@@ -202,6 +202,35 @@ def test_centralize(self, inputs, outputs):
+    (0.3, 0.54, 234),
+    (1.0, 0.56, 314),
+  )
+  def test_add_noise(self, eta, gamma, seed):
```
Could you provide more information on the behaviour being tested? Either by changing the name to e.g. `test_add_noise_has_correct_standard_deviation` or by adding a docstring with something similar (just so that there is more information if the test fails). Thanks a lot!

Rupt replied:
I've gone with `test_add_noise_has_correct_variance_scaling` since the test doesn't actually check the standard deviation, just the scale relative to eta=1, gamma=0.
Thanks for all the nice comments.

mkunesch reviewed (changes requested):
Looks great! Thank you so much for this bug fix and the associated test!
I just left a few very minor comments/nits.

mkunesch reviewed (changes requested):
Thanks a lot for making the changes! I forgot to check for `chex.all_variants` last time so I have one more comment. Otherwise looks good to me!

Rupt commented on optax/_src/transform_test.py:
```diff
@@ -202,6 +202,35 @@ def test_centralize(self, inputs, outputs):
+    (0.3, 0.54, 234),
+    (1.0, 0.56, 314),
+  )
+  def test_add_noise_has_correct_variance_scaling(self, eta, gamma, seed):
```
Added the variants.

Agreed there is nothing new the other tests can catch. Kwargs there don't work because the parameters are reused elsewhere (seed for the other transform, eta and gamma for scale), so I've just hard coded them as variables.

mkunesch reviewed (commented):
LGTM! Thanks a lot!

There is a test failure now in the (recently added) test for `noisy_sgd` in `alias_test.py` which uses `add_noise`. Do you have time to look into this? It's probably just a matter of changing the parameters so that with the slower decay of the noise it doesn't affect the test results but we should make sure this is indeed the reason.

Rupt commented:
> It's probably just a matter of changing the parameters
I've rebased and changed eta to match the previous initial variance. It survives the test with the different decay rate.
 
> but we should make sure this is indeed the reason.
The test passes now (and also passed if just removing the noisy_sgd line.

mkunesch reviewed (approved):
Great, thanks a lot for doing this!
